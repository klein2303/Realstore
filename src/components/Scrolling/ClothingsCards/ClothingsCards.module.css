.clothingSection {
    background-color: #ffffff;
    border-radius: 10px;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    height: 410px;
    width: 220px;
    transition:
        transform 0.3s ease,
        box-shadow 0.3s ease;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    margin-top: 40px;
}

.imgContainer {
    background-color: white;
    border-radius: 10px;
    height: 220px;
    width: 80%;
    transition:
        transform 0.3s ease,
        box-shadow 0.3s ease;
    overflow: hidden;
    display: flex;
    justify-content: center;
    align-items: center;
}

.clothingImage {
    max-height: 100%;
    max-width: 80%;
    padding-left: 20%;
    display: block;
}

.descriptionSpace {
    background-color: white;
    border-radius: 10px;
    height: 130px;
    width: 100%;
    transition:
        transform 0.3s ease,
        box-shadow 0.3s ease;
    justify-content: center;
    padding-top: 20px;
}

.productType {
    width: 90%;
    height: 10%;
    margin-left: 5%;
    display: flex;
}

.nameAndPrice {
    width: 90%;
    height: 70px;
    justify-content: space-between;
    align-items: center;
    flex: 1;
    display: flex;
    margin-left: 10px;
}

.productname {
    display: flex;
    align-items: start;
    text-align: left;
    margin-right: 10px;
}

.price {
    padding-right: 0;
    text-align: right;
    white-space: nowrap;
    font-weight: 200;
}

.category {
    color: #424242;
    font-weight: 200;
    font-size: 15px;
}

.h4text {
    font-family: "Georgia", sans-serif;
    color: #424242;
    font-weight: 200;
}

.productType {
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.IconSpace {
    cursor: pointer;
    display: flex;
    font-size: 20px;
    gap: 5px;
}

.IconSpaceSvg {
    font-size: 24px;
    margin-left: 5px;
}

.favorited {
    color: red;
}

@media (max-width: 800px) {
    .clothingSection {
        height: 330px;
        width: 220px;
    }
}
