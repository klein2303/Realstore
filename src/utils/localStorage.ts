// Function to store favorite items in local storage
export const setFavorites = (items: number[]): void => {
    localStorage.setItem("favorites", JSON.stringify(items));
};

// Function to get favorite items from local storage
export const getFavorites = (): number[] => {
    const storedItems = localStorage.getItem("favorites");
    return storedItems ? JSON.parse(storedItems) : [];
};

export const toggleFavorite = (productId: number): void => {
    const favorites = getFavorites();
    const isFavorite = favorites.includes(productId);

    const updatedFavorites = isFavorite ? favorites.filter((id) => id !== productId) : [...favorites, productId];

    setFavorites(updatedFavorites);
};
