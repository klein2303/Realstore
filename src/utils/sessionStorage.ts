import {Filter} from "../pages/Shopping";

export type CartItem = {productId: number; quantity: number};

// Helper function to store items in session storage
const setSessionStorageItems = <T>(key: string, items: T[]): void => {
    sessionStorage.setItem(key, JSON.stringify(items));
};

// Helper function to get items from session storage
const getSessionStorageItems = <T>(key: string): T[] => {
    const storedItems = sessionStorage.getItem(key);
    return storedItems ? JSON.parse(storedItems) : [];
};

// Cart-specific functions
export const getCart = (): CartItem[] => {
    return getSessionStorageItems<CartItem>("cart");
};

export const addToCart = (productId: number, quantity: number = 1): void => {
    const cart = getCart();
    const existingItem = cart.find((item) => item.productId === productId);

    if (existingItem) {
        existingItem.quantity += quantity; // Increment quantity
    } else {
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart);
};

export const clearCart = (): void => {
    setSessionStorageItems("cart", []);
};

export const updateCartQuantity = (productId: number, quantity: number): void => {
    const cart = getCart();
    const index = cart.findIndex((item) => item.productId === productId);

    if (index !== -1) {
        // Product exists in the cart
        if (quantity > 0) {
            cart[index].quantity = quantity; // Update quantity
        } else {
            cart.splice(index, 1); // Remove if quantity is 0
        }
    } else if (quantity > 0) {
        // Product doesn't exist in the cart, add it
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart); // Save updated cart to sessionStorage
};

// Filtered items-specific functions
export const getFilter = (): Filter[] => {
    return getSessionStorageItems<Filter>("filter");
};

export const setFilter = (filter: Filter[]): void => {
    setSessionStorageItems("filter", filter);
};

export const clearFilter = (): void => {
    setSessionStorageItems("filter", []);
};
