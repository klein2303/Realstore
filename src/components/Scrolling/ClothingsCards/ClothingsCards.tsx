import {useEffect, useState} from "react";
import styles from "./ClothingsCards.module.css";
import {IoCart, IoCartOutline} from "react-icons/io5";
import {IoMdHeartEmpty, IoMdHeart} from "react-icons/io";
import {getFavorites, toggleFavorite} from "../../../utils/localStorage";
import {Link, useLocation, useNavigate} from "react-router-dom";
import {addToCart, getCart, updateCartQuantity} from "../../../utils/sessionStorage";
import {CartItem} from "../../../utils/sessionStorage";

interface InterfaceProductCard {
    id: number;
    title: string;
    price: number;
    image: string;
    category: string;
    cart: boolean;
    favorite: boolean;
}

const ClothingsCards = ({id, title, price, category, image, cart, favorite}: InterfaceProductCard) => {
    const [isAddedToCart, setIsAddedToCart] = useState<boolean>(false);
    const router = useLocation();
    const navigate = useNavigate();

    const handleCart = () => {
        if (isAddedToCart) {
            updateCartQuantity(id, 0);
            setIsAddedToCart(false);
        } else {
            addToCart(id);
            setIsAddedToCart(true);
        }

        if (router.pathname == "/ShoppingCart") {
            navigate(0);
        }
    };

    const [isFavorited, setIsFavorited] = useState<boolean>(false);

    const handleFavorites = () => {
        toggleFavorite(id);
        setIsFavorited(!isFavorited);
        if (router.pathname == "/Favorites") {
            navigate(0);
        }
    };

    useEffect(() => {
        const currentCart = getCart().map((cart: CartItem) => cart.productId);
        if (currentCart?.includes(id)) {
            setIsAddedToCart(true);
        }

        const currentfavorites = getFavorites();
        if (currentfavorites?.includes(id)) {
            setIsFavorited(true);
        }
    }, [id]);

    return (
        <article className={styles.clothingSection} role="region" aria-label={`Product card for ${title}`}>
            <figure className={styles.imgContainer}>
                <Link to={`/Clothe/${id}`}>
                    <img src={image} className={styles.clothingImage} alt={`Image of ${title}`}></img>
                </Link>
            </figure>
            <section className={styles.descriptionSpace}>
                <header className={styles.productType}>
                    <h5 className={styles.category}>{category}</h5>
                    <section className={styles.IconSpace} aria-label="Product Actions">
                        {cart && (
                            <span
                                onClick={handleCart}
                                onKeyDown={(e) => e.key === "Enter" && handleCart()}
                                aria-label={isAddedToCart ? "Remove from cart" : "Add to cart"}
                                role="button">
                                {isAddedToCart ? <IoCart tabIndex={0} /> : <IoCartOutline tabIndex={0} />}
                            </span>
                        )}
                        {favorite && (
                            <span
                                onClick={handleFavorites}
                                onKeyDown={(e) => e.key === "Enter" && handleFavorites()}
                                aria-label={isFavorited ? "Remove from favorites" : "Add to favorites"}
                                role="button">
                                {isFavorited ? (
                                    <IoMdHeart className={styles.favorited} tabIndex={0} />
                                ) : (
                                    <IoMdHeartEmpty tabIndex={0} />
                                )}
                            </span>
                        )}
                    </section>
                </header>
                <footer className={styles.nameAndPrice}>
                    <span className={styles.productname}>
                        <h4 className={styles.h4text}>{title}</h4>
                    </span>
                    <span className={styles.price}>
                        <h4 className={styles.h4text}>$ {price}</h4>
                    </span>
                </footer>
            </section>
        </article>
    );
};

export default ClothingsCards;
