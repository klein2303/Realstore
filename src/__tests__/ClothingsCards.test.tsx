import {render, screen, fireEvent} from "@testing-library/react";
import {describe, it, expect, vi, beforeEach} from "vitest";
import '@testing-library/jest-dom';
import ClothingsCards from "../components/Scrolling/ClothingsCards/ClothingsCards";
import {MemoryRouter} from "react-router-dom";
import {getCart, addToCart, updateCartQuantity} from "../utils/sessionStorage";
import {getFavorites, toggleFavorite} from "../utils/localStorage";

// Mock sessionStorage and localStorage functions
vi.mock("../utils/sessionStorage", () => ({
    getCart: vi.fn(),
    addToCart: vi.fn(),
    updateCartQuantity: vi.fn(),
}));

vi.mock("../utils/localStorage", () => ({
    getFavorites: vi.fn(),
    toggleFavorite: vi.fn(),
}));

describe("ClothingsCards Component", () => {
    const mockProduct = {
        id: 1,
        title: "Test Product",
        price: 100,
        image: "test-image.jpg",
        category: "Clothing",
        cart: true,
        favorite: true,
    };

    beforeEach(() => {
        vi.resetAllMocks();
        // Default mocks to avoid undefined errors
        (getCart as jest.Mock).mockReturnValue([]);
        (getFavorites as jest.Mock).mockReturnValue([]);
    });

    it("renders the product card with correct details", () => {
        render(
            <MemoryRouter>
                <ClothingsCards {...mockProduct} />
            </MemoryRouter>,
        );

        // Check if the product card renders with correct details
        const productCard = screen.getByRole("region", {name: /product card for test product/i});
        expect(productCard).toBeInTheDocument();

        const productImage = screen.getByAltText(/image of test product/i);
        expect(productImage).toHaveAttribute("src", "test-image.jpg");

        expect(screen.getByText(/test product/i)).toBeInTheDocument();
        expect(screen.getByText(/\$ 100/i)).toBeInTheDocument();
        expect(screen.getByText(/clothing/i)).toBeInTheDocument();
    });

    it("handles adding and removing items from the cart", () => {
        (getCart as jest.Mock).mockReturnValue([{productId: 1}]);

        render(
            <MemoryRouter>
                <ClothingsCards {...mockProduct} />
            </MemoryRouter>,
        );

        const cartIcon = screen.getByRole("button", {name: /cart/i});

        // Simulate removing from the cart
        fireEvent.click(cartIcon);
        expect(updateCartQuantity).toHaveBeenCalledWith(1, 0);

        // Simulate adding to the cart
        fireEvent.click(cartIcon);
        expect(addToCart).toHaveBeenCalledWith(1);
    });

    it("handles adding and removing items from favorites", () => {
        (getFavorites as jest.Mock).mockReturnValue([1]);

        render(
            <MemoryRouter>
                <ClothingsCards {...mockProduct} />
            </MemoryRouter>,
        );

        const favoriteIcon = screen.getByRole("button", {name: /favorite/i});

        // Simulate removing from favorites
        fireEvent.click(favoriteIcon);
        expect(toggleFavorite).toHaveBeenCalledWith(1);

        // Simulate adding to favorites
        fireEvent.click(favoriteIcon);
        expect(toggleFavorite).toHaveBeenCalledWith(1);
    });

    it("matches the snapshot", () => {
        const {container} = render(
            <MemoryRouter>
                <ClothingsCards {...mockProduct} />
            </MemoryRouter>,
        );

        expect(container).toMatchSnapshot();
    });

    it("initializes states based on localStorage and sessionStorage", () => {
        (getCart as jest.Mock).mockReturnValue([{productId: 1}]);
        (getFavorites as jest.Mock).mockReturnValue([1]);

        render(
            <MemoryRouter>
                <ClothingsCards {...mockProduct} />
            </MemoryRouter>,
        );

        // Check that initial cart state is correct
        const cartIcon = screen.getByRole("button", {name: /cart/i});
        expect(cartIcon).toBeInTheDocument();

        // Check that initial favorites state is correct
        const favoriteIcon = screen.getByRole("button", {name: /favorite/i});
        expect(favoriteIcon).toBeInTheDocument();
    });
});
