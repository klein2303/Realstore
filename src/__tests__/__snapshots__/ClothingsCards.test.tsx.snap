// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ClothingsCards Component > matches the snapshot 1`] = `
<div>
  <article
    aria-label="Product card for Test Product"
    class="_clothingSection_47ab48"
    role="region"
  >
    <figure
      class="_imgContainer_47ab48"
    >
      <a
        href="/Clothe/1"
      >
        <img
          alt="Image of Test Product"
          class="_clothingImage_47ab48"
          src="test-image.jpg"
        />
      </a>
    </figure>
    <section
      class="_descriptionSpace_47ab48"
    >
      <header
        class="_productType_47ab48"
      >
        <h5
          class="_category_47ab48"
        >
          Clothing
        </h5>
        <section
          aria-label="Product Actions"
          class="_IconSpace_47ab48"
        >
          <span
            aria-label="Add to cart"
            role="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              tabindex="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="176"
                cy="416"
                fill="none"
                r="16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
              <circle
                cx="400"
                cy="416"
                fill="none"
                r="16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
              <path
                d="M48 80h64l48 272h256"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
              <path
                d="M160 288h249.44a8 8 0 0 0 7.85-6.43l28.8-144a8 8 0 0 0-7.85-9.57H128"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
            </svg>
          </span>
          <span
            aria-label="Add to favorites"
            role="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              tabindex="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M349.6 64c-36.4 0-70.7 16.7-93.6 43.9C233.1 80.7 198.8 64 162.4 64 97.9 64 48 114.2 48 179.1c0 79.5 70.7 143.3 177.8 241.7L256 448l30.2-27.2C393.3 322.4 464 258.6 464 179.1 464 114.2 414.1 64 349.6 64zm-80.8 329.3l-4.2 3.9-8.6 7.8-8.6-7.8-4.2-3.9c-50.4-46.3-94-86.3-122.7-122-28-34.7-40.4-63.1-40.4-92.2 0-22.9 8.4-43.9 23.7-59.3 15.2-15.4 36-23.8 58.6-23.8 26.1 0 52 12.2 69.1 32.5l24.5 29.1 24.5-29.1c17.1-20.4 43-32.5 69.1-32.5 22.6 0 43.4 8.4 58.7 23.8 15.3 15.4 23.7 36.5 23.7 59.3 0 29-12.5 57.5-40.4 92.2-28.8 35.7-72.3 75.7-122.8 122z"
              />
            </svg>
          </span>
        </section>
      </header>
      <footer
        class="_nameAndPrice_47ab48"
      >
        <span
          class="_productname_47ab48"
        >
          <h4
            class="_h4text_47ab48"
          >
            Test Product
          </h4>
        </span>
        <span
          class="_price_47ab48"
        >
          <h4
            class="_h4text_47ab48"
          >
            $ 
            100
          </h4>
        </span>
      </footer>
    </section>
  </article>
</div>
`;
